x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Realms.java

```diff
@@ -100,8 +104,11 @@ public Realms(
             .distinct()
             .collect(Collectors.toMap(DomainConfig::name, Function.identity()));
         final List<RealmConfig> realmConfigs = buildRealmConfigs();
+
+        // initRealms will add default file and native realm config if they are not explicitly configured
         final List<Realm> initialRealms = initRealms(realmConfigs);
-        configureRealmRef(initialRealms, realmConfigs, realmToDomainConfig);
+        realmRefs = calculateRealmRefs(realmConfigs, realmToDomainConfig);
+        initialRealms.forEach(realm -> realm.initRealmRef(realmRefs.get(new RealmConfig.RealmIdentifier(realm.type(), realm.name()))));
```

Comment:
nit: `initialRealms.forEach(realm -> realm.initRealmRef(realmRefs))` , I prefer it because the current code is verbose and I feel it adds to the confusion.

Reply:
Updated as suggested. Thanks!